Write the tasks-and-scripts guide (#2851)
NomicFoundation/hardhat: docs/src/content/hardhat-runner/docs/guides/tasks-and-scripts.md

# Writing tasks and scripts

> At its core, Hardhat is a task runner. This means that you can define and run tasks to automate things. Hardhat also comes with some built-in tasks, like `compile` and `test`, but others can be added via plugins or just by defining new tasks in your configuration file.

Reviewer: ```suggestion
At its core, Hardhat is a task runner that allows you to automate your development workflow. It comes with some built-in tasks, like `compile` and `test`, but you can define new ones in your configuration file, or import them via plugins.
```

This guide assumes you have initialized a sample project. If not, check [this guide](./project-setup.md).

### A simple task

> Let's write a very simple task that just prints the list of available accounts. Add this to your config:

Reviewer: ```suggestion
Let's write a very simple task that prints the list of available accounts. 

Start by adding this to your config:
```

```js
task("accounts", "Prints the list of accounts", async (taskArgs, hre) => {
  const accounts = await hre.ethers.getSigners();

  for (const account of accounts) {
    console.log(account.address);
  }
});
```

> And run the new task:

Reviewer: ```suggestion
Now you can run the new task:
```

```
npx hardhat accounts
```

> Let's see how this works. We are using the `task` function to define our new task. The first argument is the name of the task and it's what we use in the command line to run it. The second argument is the description of the task, which is printed when you use `npx hardhat help`.

Reviewer: ```suggestion
We are using the `task` function to define our new task. Its first argument is the name of the task and it's what we use in the command line to run it. The second argument is the description of the task, which is printed when you use `npx hardhat help`.
```

> The third argument is the definition of the task itself. It's a function that receives two arguments: the arguments for the task (we don't have any yet) and the [Hardhat Runtime Environment](../advanced/hardhat-runtime-environment.md) (HRE), containing all the functionality of Hardhat and of any enabled plugins. The body of this function is what gets executed when you run the task. Here we are using `ethers.getSigners()` to obtain all the configured accounts and then, for each one, we print its address.

Reviewer: ```suggestion
The third argument is the definition of the task itself. It's an async function that receives two arguments: the arguments for the task (we don't have any yet!) and the [Hardhat Runtime Environment](../advanced/hardhat-runtime-environment.md) (HRE), containing all the functionality of Hardhat and its plugins. The body of this function is what gets executed when you run the task. Here we are using `ethers.getSigners()` to obtain all the configured accounts and then preach each of their addresses.
```

# Writing tasks and scripts

> At its core, Hardhat is a task runner that allows you to automate your development workflow. It comes with some built-in tasks, like `compile` and `test`, but you can customize it.

Reviewer 2: ```suggestion
At its core, Hardhat is a task runner that allows you to automate your development workflow. It comes with some built-in tasks, like `compile` and `test`, but you can add your own custom tasks as well.
```

> In this guide will show you how to extend Hardhat's functionality using tasks and scripts. It assumes you have initialized a sample project. If you haven't done it, please read [this guide](./project-setup.md) first.

Reviewer 2: ```suggestion
This guide will show you how to extend Hardhat's functionality using tasks and scripts. It assumes you have initialized a sample project. If you haven't done it, please read [this guide](./project-setup.md) first.
```

## Writing Hardhat Tasks

Let's write a very simple task that prints the list of available accounts, and explore how it works.

> Please, copy this task definition and paste it into your config:

Reviewer 2: ```suggestion
Copy this task definition and paste it into your hardhat config file:
```
